tango/executors/multicore_executor.py

```diff
@@ -0,0 +1,250 @@
+        return errors
+
+    def _sync_step_states(self, step_graph: StepGraph) -> Dict[str, StepState]:
+        """
+        Update the StepState info.
+        Although, this is not really elegant. The issue is as follows: The main multicore executor process
+        queues a step, and starts step execution in a different process. If we try to read the StepState
+        before that process has had time to update the StepState, the Workspace will throw the out of sync
+        error (IOError: process should be running but it's considered incomplete...).
+
+        Hence, we try to read a few times, so that the child process has time to update the step's state.
+        """
+
+        import time
+
+        attempts = 0
+        while attempts < self._num_tries_to_sync_states:
+            attempts += 1
+            try:
+                step_states = {step.name: self._get_state(step) for step in step_graph.values()}
+                break
+            except IOError:
+                if attempts == self._num_tries_to_sync_states:
+                    raise
+                step_states = {}
+                time.sleep(self._wait_seconds_to_sync_states)
+        return step_states
+
+    def execute_step_graph(self, step_graph: StepGraph):
+        """
+        Execute a :class:`tango.step_graph.StepGraph`.
+        """
+
+        self._reset_execution()
+
+        from tempfile import NamedTemporaryFile
+
+        # Creates a temporary file in which to store the config. This is passed as a command line
+        # argument to child step processes.
+        with NamedTemporaryFile(prefix="step-graph-to-file-run", suffix=".jsonnet") as file_ref:
+            step_graph.to_file(file_ref.name)
+            assert os.path.exists(file_ref.name)
+
+            step_states = self._sync_step_states(step_graph)
+
+            while self._has_incomplete_steps(step_graph, step_states):
+                # Cleanup previously running steps.
+                self._update_running_steps(step_states)
+
+                # Get steps that are ready to run.
+                to_run = self._get_steps_to_run(step_graph, step_states)
+                if to_run:
+                    logger.debug(f"Steps ready to run: {to_run}")
+
+                for step_name in to_run:
+                    self._queue_step(step_name)
+
+                # Begin processes for any queued steps (if not enough processes are already running).
+                while len(self._queued_steps) > 0 and len(self._running) < self.parallelism:
+                    self._try_to_execute_next_step(config_path=file_ref.name)
+
+                # Re-sync the StepState info.
+                step_states = self._sync_step_states(step_graph)
+
+        assert not self._running and not self._queued_steps
+
+    @property
+    def failed_steps(self) -> Set[str]:
+        return self._failed
+
+    def _has_incomplete_steps(
+        self, step_graph: StepGraph, step_states: Dict[str, StepState]
+    ) -> bool:
```

Comment:
I make a distinction because `_has_incomplete_steps` needs to wait for steps that are currently running as well, whereas, `_get_steps_to_run` will only give me the steps that can be run immediately (dependencies are satisfied, etc.).